packages/in-app-notification-dispatcher/src/handlers/eservices/handleEserviceNewVersionApprovedRejectedToDelegate.ts

```
@@ -0,0 +1,86 @@
+  eserviceV2Msg: EServiceV2 | undefined,
+  descriptorId: DescriptorId,
+  logger: Logger,
+  readModelService: ReadModelServiceSQL,
+  eventType: EserviceNewVersionApprovedRejectedToDelegateEventType
+): Promise<NewNotification[]> {
+  if (!eserviceV2Msg) {
+    throw missingKafkaMessageDataError("eservice", eventType);
+  }
+  logger.info(
+    `Handle eservice new version approved/rejected in-app notification for eservice ${eserviceV2Msg.id}`
+  );
+
+  const eservice = fromEServiceV2(eserviceV2Msg);
+
+  const producerDelegation = await readModelService.getActiveProducerDelegation(
+    eservice.id,
+    eservice.producerId
+  );
+
+  if (!producerDelegation) {
+    throw activeProducerDelegationNotFound(eservice.id);
+  }
+
+  const usersWithNotifications =
+    await readModelService.getTenantUsersWithNotificationEnabled(
+      [producerDelegation.delegateId],
+      "eserviceNewVersionApprovedRejectedToDelegate"
+    );
+
+  if (usersWithNotifications.length === 0) {
+    logger.info(
+      `No users with notifications enabled for eservice ${eserviceV2Msg.id}`
+    );
+    return [];
+  }
+
+  const delegator = await retrieveTenant(eservice.producerId, readModelService);
+
+  const rejectionReason = getLastRejectionReason(eservice, descriptorId);
+
+  const body = inAppTemplates.eserviceNewVersionApprovedRejectedToDelegate(
+    delegator.name,
+    eservice.name,
+    eventType,
+    rejectionReason
+  );
+
+  return usersWithNotifications.map(({ userId, tenantId }) => ({
+    userId,
+    tenantId,
+    body,
+    notificationType: "eserviceNewVersionApprovedRejectedToDelegate",
+    entityId: descriptorId,
+  }));
+}
+
+const getLastRejectionReason = (
+  eservice: EService,
+  descriptorId: DescriptorId
+): string | undefined => {
+  const descriptor = eservice.descriptors.find(
+    (descriptor) => descriptor.id === descriptorId
+  );
+  return descriptor?.rejectionReasons?.[descriptor.rejectionReasons.length - 1]
+    .rejectionReason;
```

Please double-check that `rejectionsReasons` is actually a sorted array, inside the function `aggregateDescriptor` (in `readmodel` package). Let me know if you need more context.

---

Good point, thanks. I think the code should be correct because it reads an event payload, which should preserve the order in which the array was created in https://github.com/pagopa/interop-be-monorepo/blob/0de498892541c7a946a02c0183fba3ce6f66925e/packages/catalog-process/src/services/catalogService.ts#L2697-L2700

In contrast, when writing/reading from the readmodel, the order would not necessarily be preserved (since it's a separate table and not a Postgres array).

But, at least to make the code more robust, so it wouldn't break if we end up using an eservice read from the readmodel instead of from the event, I guess it's safer to take the most recent `rejectedAt` which is what the frontend does. Does that make sense to you?

---

I agree, it seems the last one is for sure the most recent, whereas the previous ones might not be sorted.
Up to you if you prefer to update the function to find the most recent `rejectedAt` for robustness.
Anyway, I have also created the task PIN-7557 to fix the aggregator in `readmodel`.